### PR TITLE
Resolved bug/testcase specific karma error

### LINKED ISSUE #2456: bugs can't be marked as fixed
In https://bodhi.fedoraproject.org/updates/FEDORA-2018-1a6e6196b9 I tried to mark #1580530 as fixed by giving it +1 feedback (directly for that bug). But only generic karma was shown in my [comment](https://bodhi.fedoraproject.org/updates/ansible-2.5.5-2.fc28#comment-800652), not the specific bug report karma. Also in Bugs tab the feedback was not shown. I tried again (this time without giving generic karma), and [again](https://bodhi.fedoraproject.org/updates/ansible-2.5.5-2.fc28#comment-800653) it was ignored.

It seems that giving karma for particular bugs is broken atm.

### PR DESCRIPTION
**Pull request by ghost**
fixes #2456, #2189, #2130

Signed-off-by: Vismay Golwala <vgolwala@redhat.com>

### CODE REVIEW

File: bodhi/server/validators.py [hunk 1]
```diff
@@ -989,6 +989,39 @@ def validate_comment_id(request, **kwargs):
...
+
+    Args:
+        request (pyramid.util.Request): The current request.
+        kwargs (dict): The kwargs of the related service definition. Unused.
+    """
+    # In-method import used because there is mutual referencing between
+    # 'bodhi.server.schemas' and 'bodhi.server.validators'
+    from bodhi.server.schemas import SaveCommentSchema
+
+    content = str(request.body)
```
Comment (cverna) [inline bodhi/server/validators.py]: I think here you could use the [request.json ](https://docs.pylonsproject.org/projects/pyramid/en/latest/api/request.html#pyramid.request.Request.json)attribute that will give you directly a dictionary to work with.

File: bodhi/server/validators.py [hunk 2]
```diff
@@ -989,6 +989,39 @@ def validate_comment_id(request, **kwargs):
...
+
+    if content:
+        content = eval(content)
+
+        if type(content) is dict:
+            if 'bug_feedback' not in content and 'testcase_feedback' not in content:
+                content = SaveCommentSchema().unflatten(content)
+            content = str(content)
+
+    if type(content) is str:
```
Comment (cverna) [inline bodhi/server/validators.py]: You should check the pyramid escape function for this. See https://docs.pylonsproject.org/projects/pyramid/en/latest/_modules/html.html#escape

File: bodhi/server/validators.py [hunk 3]
```diff
@@ -989,6 +989,39 @@ def validate_comment_id(request, **kwargs):
...
+        request.body = content.replace("'", "\"")
+    if type(content) is bytes:
+        request.body = content.replace(b"'", b"\"")
```
Comment (cverna) [inline bodhi/server/validators.py]: What is the use case of receiving bytes in the body ?

**Review by cverna — CHANGES_REQUESTED** [review]

File: bodhi/server/validators.py (hunk 1 above)
Comment (ghost) [reply bodhi/server/validators.py]: But that won't be set in test environment. So should we set that also in testing or make a decision based logic of whether parameters are set?

Comment (cverna) [reply bodhi/server/validators.py]: in the test instead of setting `request.body` you should be able to set `request.json` instead.

Comment (ghost) [reply bodhi/server/validators.py]: ```request.json``` stores every key-value pair in ```unicode``` format. However, the ```colander``` parser requires it to be ```str```. So I can either convert each key-value of ```request.json``` into ```str``` or I can use the method above. Which one would be more efficient according to you?

File: bodhi/server/validators.py (hunk 3 above)
Comment (ghost) [reply bodhi/server/validators.py]: Testing using python 3, it sets the ```request.body``` as ```bytes``` and in that case, force converting into ```str``` causes the following error:
```TypeError: You can only set Request.body to bytes (not <class 'str'>)```

That's why, I kept separate logics for python 2 and 3.

File: bodhi/server/validators.py (hunk 2 above)
Comment (ghost) [reply bodhi/server/validators.py]: In the line you are suggesting, I am not escaping quotes - but instead converting single quotes to double quotes because that is the format colander parses string in.

However, it might be good to actually escape the ```text``` field of request to filter HTML characters.

Comment (cverna) [reply bodhi/server/validators.py]: ok I need to look at colander in more details :small_airplane:

**Comment by cverna** [comment]
@vismay-golwala, I gave a quick look at [colander](https://docs.pylonsproject.org/projects/colander/en/latest/basics.html)  and I don't think we should use a validator to format the request body the way colander is validating it.

I think we should look at the colander deserialize or preparer to format the data before they get validated.

Also I tried running this code using Python 3 and it is not working. I suspect the `str()` calls are not doing what you expect in Python 3.

**Comment by ghost** [comment]
@cverna , so the process of execution is somewhat like this:
```submit_feedback (request object) -> ColanderSchema (deserialize) -> Preparer -> Validator -> rest_of_code```


```deserializing``` needs the data to be 'unflattened'. So I think the way to achieve it is we filter the request object before it goes to ```ColanderSchema```. This is not data validation, but instead, request validation which just converts it into an appropriate format for deserialization. 


If we go for ```Preparer```, it gets executed after deserialization. So, if the data was not formatted before, it would cause ```bug_feedback``` or ```testcase_feedback``` to be empty when it reaches ```Preparer```.


Regarding Python 3 issue, can you please tell me how to run the server instance using Py3 instead of 2?

File: .pytest_cache/v/cache/lastfailed [hunk 4]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (cverna) [inline .pytest_cache/v/cache/lastfailed]: this file should not be part of the commit. You could add `.pytest_cache` to the `.gitignore` file

File: bodhi/server/schemas.py [hunk 5]
```diff
@@ -145,6 +145,11 @@ class TestcaseFeedbacks(colander.SequenceSchema):
 class SaveCommentSchema(CSRFProtectedSchema, colander.MappingSchema):
     """An API schema for bodhi.server.services.comments.new_comment()."""
 
+    def deserialize(self, cstruct):
+        """Unflatten comment before parsing into Schema."""
+        appstruct = SaveCommentSchema().unflatten(cstruct)
+        return super(SaveCommentSchema, self).deserialize(appstruct)
+
```
Comment (cverna) [inline bodhi/server/schemas.py:152]: Nice :+1:

**Review by cverna — CHANGES_REQUESTED** [review]
Code looks good, there is just one file that should not be included in the commit

**Review by cverna — APPROVED** [review]
+1
